flink-streaming-java/src/main/java/org/apache/flink/streaming/runtime/tasks/AsyncCheckpointRunnable.java

```diff
@@ -197,7 +199,8 @@ private void handleExecutionException(Exception e) {
 				// We only report the exception for the original cause of fail and cleanup.
 				// Otherwise this followup exception could race the original exception in failing the task.
 				try {
-					taskEnvironment.declineCheckpoint(checkpointMetaData.getCheckpointId(), checkpointException);
+					taskEnvironment.declineCheckpoint(checkpointMetaData.getCheckpointId(),
+							new CheckpointException(CheckpointFailureReason.EXCEPTION, checkpointException));
```

Comment:
I have two concerns for this change:

- I saw the motivation from the respective jira description that it wants to resolve the misleading `CheckpointFailureReason.JOB_FAILURE`. But when I traced the codes, I found the `JOB_FAILURE` was used in the below region of `CheckpointCoordinator#receiveDeclineMessage`:

```
final CheckpointException checkpointException;
if (message.getReason() == null) {
		checkpointException = new CheckpointException(CheckpointFailureReason.CHECKPOINT_DECLINED);
} else {
		checkpointException = getCheckpointException(CheckpointFailureReason.JOB_FAILURE, message.getReason());
}
```

So I guess whatever which exception is wrapped here, it would be all regarded as `JOB_FAILURE` unless `null` reason. Or I misunderstood something else?

- The semantic seems not correct when we use `CheckpointFailureReason.EXCEPTION` here, because `CheckpointFailureReason#preFlight` will be true for this exception. But we know that if the exception happens during async checkpoint process, the semantic of `#preFlight` should be false.

Another tiny comment for the formatting is to either make every argument in separate line or both in the same line. :)

Reply:
> CheckpointFailureReason.JOB_FAILURE is the default value of method getCheckpointException, you can check the implementation for more details.

Yeah, you are right. 

> You're right about the semantic. In our internal version we add a new CheckpointFailureReason value:

I like the `CHECKPOINT_ASYNC_EXCEPTION` way. Go ahead!

BTW, how about supplementing somehow unit test for the changes if possible?  
Maybe construct `AsyncCheckpointRunnable`  like did in `LocalStateForwardingTest`  and then throw an exception during `run`. Then we can override the implementation of `StreamMockEnvironment#declineCheckpoint` to verify the exception argument `instanceof CheckpointException` and `preFlight = false`. Or there are some other easier way I did not think through.

Reply:
How about adding two tests? 

1. Verify the exception thrown from `AsyncCheckpointRunnable`.
2. Verify the `CheckpointReason` on `CheckpointCoordinator`'s side.

Reply:
I think the first one is necessary since the type of thrown exception is changed in this PR.
The second one is optional since we do not change any logics there. But if you want to supplement some previously missing unit tests for the `CheckpointCoordinator` component, I am also fine with it. :)